### PR TITLE
add endpoint to sideload snaps

### PR DESCRIPTION
**Pull request by st3v3nmw**
Follow up on #3

### CODE REVIEW

File: snap_http/api.py [hunk 1]
```diff
@@ -109,6 +109,52 @@ def install_all(names: List[str]) -> SnapdResponse:
...
+
+    if jailmode:
+        data["jailmode"] = jailmode
+
+    if system_restart_immediate:
+        data["system-restart-immediate"] = system_restart_immediate
+
+    files = []
+    for file_path in file_paths:
+        files.append(FileUpload(name="snap", path=file_path))
```
Comment (Perfect5th) [inline snap_http/api.py:153]: Is there any concern when we have multiple `FileUpload` instances with the same `name`?

File: tests/integration/test_api.py [hunk 2]
```diff
@@ -54,3 +54,27 @@ def test_remove_snap(test_snap):
     response = wait_for(snap_http.remove)("test-snap")
     assert response.status_code == 202
     assert is_snap_installed("test-snap") is False
+
+
+def test_sideload_snap(local_test_snap_path):
```
Comment (Perfect5th) [inline tests/integration/test_api.py:59]: please add a test with more than a single `file_path` so we get the n > 1 case covered.

I guess that means you need to have two test snaps to use? Or can you use the same one twice?

File: tests/integration/test_api.py [hunk 3]
```diff
@@ -54,3 +54,27 @@ def test_remove_snap(test_snap):
...
+    """Test sideloading a snap."""
+    assert is_snap_installed("test-snap") is False
+
+    response = wait_for(snap_http.sideload)(
+        file_paths=[local_test_snap_path],
+        devmode=True,
```
Comment (Perfect5th) [inline tests/integration/test_api.py:65]: up to you if you think we need further integration tests with any of the other `sideload` boolean flags.

File: tests/unit/test_types.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            "--c747ef5f-cef1-40a7-86e5-12ca0aff6cd5--\r\n"
+        ).encode()
+        assert body.serialized == expected
+        assert body.content_length == len(expected)
+
+
+def test_reading_file_uploads():
+    """Test reading of file uploads."""
+    with tempfile.NamedTemporaryFile() as tmp:
+        tmp.write(b"the answer is 42")
```
Comment (Perfect5th) [inline tests/unit/test_types.py:98]: but what's the question? :wink:

**Review by Perfect5th — CHANGES_REQUESTED** [review]
Mostly minor stuff. Looks pretty good.

File: snap_http/api.py (hunk 1 above)
Comment (st3v3nmw) [reply snap_http/api.py:153]: I don't think so, the API does it that way:

> Snaps can be sideloaded by passing the snap content in a multipart/form-data request with one or more files named “snap”.

```
POST /v2/snaps HTTP/1.1
Host:
Content-Type: multipart/form-data; boundary=foo
Content-Length: 20638

--foo
Content-Disposition: form-data; name="devmode"

true
--foo
Content-Disposition: form-data; name="snap"; filename="hello-world_27.snap"

<20480 bytes of snap file data>
--foo
Content-Disposition: form-data; name="snap"; filename="other-snap_1.snap"

<20480 bytes of snap file data>
--foo
```

https://snapcraft.io/docs/snapd-api#heading--snaps

File: snap_http/api.py [hunk 5]
```diff
@@ -109,6 +109,52 @@ def install_all(names: List[str]) -> SnapdResponse:
...
+    if devmode:
+        data["devmode"] = devmode
+
+    if jailmode:
+        data["jailmode"] = jailmode
+
+    if system_restart_immediate:
+        data["system-restart-immediate"] = system_restart_immediate
+
+    files = []
```
[the remark this answers is not in the document]
Comment (st3v3nmw) [reply snap_http/api.py]: switched to list comprehension

**Review by Perfect5th — APPROVED** [review]

**Pull request merged**